src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs

```diff
@@ -275,10 +279,98 @@ Task IInterleavingContext.CallActorAsync(ActorId actorId, string operationName,
         }
 
         /// <inheritdoc/>
-        Task<IDisposable> IInterleavingContext.LockAsync(params ActorId[] actors)
+        async Task<IDisposable> IInterleavingContext.LockAsync(params ActorId[] actors)
         {
             this.ThrowIfInvalidAccess();
-            throw new NotImplementedException(); // TODO
+            if (this.ContextLocks != null)
+            {
+                throw new LockingRulesViolationException("Cannot acquire more locks when already holding some locks.");
+            }
+
+            if (actors == null || actors.Length == 0)
+            {
+                throw new ArgumentException("The list of actors to lock must not be null or empty.", nameof(actors));
+            }
+
+            // acquire the locks in a globally fixed order to avoid deadlocks
+            Array.Sort(actors);
+
+            // remove duplicates if necessary. Probably quite rare, so no need to optimize more.
+            for (int i = 0; i < actors.Length - 1; i++)
+            {
+                if (actors[i].Equals(actors[i + 1]))
+                {
+                    actors = actors.Distinct().ToArray();
+                    break;
+                }
+            }
+
+            // use a deterministically replayable unique ID for this lock request, and to receive the response
+            var lockRequestId = this.NewGuid();
+
+            // send lock request to first actor in the lock set, who will forward it to the next, etc.
+            var target = new OrchestrationInstance() { InstanceId = ActorId.GetSchedulerIdFromActorId(actors[0]) };
```

Comment:
Ok. Basically, all the actors in actor[] need to be locked, but to avoid deadlock, the locks have to be acquired sequentially, in order. So, the requestor sends the lock request to the first actor; when the first lock is granted by the first actor, the first actor forwards the lock request to the second actor, and so on; after the last actor grants the last lock, a response is sent back to the original requestor.